src/backend/dx11/shaders/copy.hlsl

```diff
@@ -1,15 +1,200 @@
-cbuffer BufferImageCopy : register(b0) {
-    uint2 BufferSize;
-    uint2 ImageOffset;
+struct BufferCopy {
+    uint4 SrcDst;
 };
 
-StructuredBuffer<uint> CopySrc : register(t0);
-RWTexture2D<uint> CopyDst : register(u0);
+struct ImageCopy {
+    uint4 Src;
+    uint4 Dst;
+};
+
+struct BufferImageCopy {
+    // x=offset, yz=size
+    uint4 BufferVars;
+    uint4 ImageOffset;
+    uint4 ImageExtent;
+};
+
+cbuffer CopyConstants : register(b0) {
+    BufferCopy BufferCopies;
+    ImageCopy ImageCopies;
+    BufferImageCopy BufferImageCopies;
+};
+
+uint2 GetImageDst(uint3 dispatch_thread_id)
+{
+    return BufferImageCopies.ImageOffset.xy + dispatch_thread_id.xy;
+}
+
+uint2 GetImageSrc(uint3 dispatch_thread_id)
+{
+    return BufferImageCopies.ImageOffset.xy + dispatch_thread_id.xy;
+}
+
+uint GetBufferDst(uint3 dispatch_thread_id)
+{
+    return BufferImageCopies.BufferVars.x + dispatch_thread_id.x + dispatch_thread_id.y * BufferImageCopies.BufferVars.y;
+}
+
+uint GetBufferSrc(uint3 dispatch_thread_id)
+{
+    return BufferImageCopies.BufferVars.x + dispatch_thread_id.x + dispatch_thread_id.y * BufferImageCopies.BufferVars.y;
+}
+
+uint Uint4ToUint(uint4 data)
+{
+    data.x = min(data.x, 0x000000ff);
+    data.y = min(data.y, 0x000000ff);
+    data.z = min(data.z, 0x000000ff);
+    data.w = min(data.w, 0x000000ff);
+
+    uint output = (data.x        |
+                  (data.y << 8)  |
+                  (data.z << 16) |
+                  (data.w << 24));
+
+    return output;
+}
+
+uint4 UintToUint4(uint data)
+{
+    return uint4((data & 0xff000000) >> 24, (data & 0xff0000) >> 16, (data & 0xff00) >> 8, data & 0xff);
+}
+
+uint2 UintToUint2(uint data)
+{
+    return uint2((data >> 16) & 0x0000ffff, data & 0x0000ffff);
+}
+
+uint Uint2ToUint(uint2 data)
+{
+    data.x = min(data.x, 0x0000ffff);
+    data.y = min(data.y, 0x0000ffff);
+
+    uint output = (data.x         |
+                  (data.y << 16));
+
+    return output;
+}
+
+// Buffers are always R32-aligned
+StructuredBuffer<uint> BufferCopySrc : register(t0);
+RWBuffer<uint>         BufferCopyDst: register(u0);
+
+// R32
+Texture2D<uint>    ImageCopySrcR32 : register(t0);
```

Comment:
we should totally share these definitions between different bit sizes (R16/R32/R8/etc)